### DISCUSSION TITLE
Permission denied to github-actions[bot] even though PAT has permission

### DISCUSSION THREAD

**Discussion by nickderobertis** [community/community#26694] [Actions]
<p>Hey all, I have <a href="https://github.com/nickderobertis/flexlate-dev-semantic-release-example/runs/6239877225?check_suite_focus=true" rel="noopener nofollow ugc">an action</a> where I am trying to push a change to the same repo. I am authenticating via personal access token (PAT).</p>
<p>For some reason on this repo, the action fails with:</p>
<pre><code class="lang-auto">Permission to nickderobertis/flexlate-dev-semantic-release-example.git denied to github-actions[bot]
</code></pre>
<p>I added some debugging to show the authentication state of the PAT. <code>gh auth status</code> shows I am properly logged in and a curl to Github’s API shows my PAT has the following permissions:</p>
<pre><code class="lang-auto">admin:org
admin:org_hook
admin:repo_hook
delete:packages
delete_repo
gist
read:user
repo
user:email
workflow
write:permissions
</code></pre>
<p>Also <code>ls-remote</code> is working fine on the repo so it seems to be a write access problem.</p>
<p>Meanwhile, on other repos, I have been able <a href="https://github.com/nickderobertis/flexlate-after-main-merge-action/runs/6239908757?check_suite_focus=true#step:3:8" rel="noopener nofollow ugc">use the same code just fine</a>. I previously had provided the same PAT to both repos, but I have also tried creating a new PAT and adding to the affected repo with no change in behavior.</p>
<p>I am guessing there must be something different about this repo that prevents access, but I’m not sure what it could be. I can’t find any difference in the settings.</p>
<p>Has anyone encountered this or have ideas on what to check? Thanks!</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>I notice you don’t set the token in <code>actions/checkout</code>, and don’t opt-out of configuring the <code>Authorization</code> header with it either. The result is that your PAT is effectively ignored, the push works if the default <code>GITHUB_TOKEN</code> is allowed to push.</p>
<p>I assume either that’s the difference between your repositories, or the one with the issue has a branch branch protection rule that gets in the way.</p>

**Comment by nickderobertis** [comment]
<p>Thank you for pointing that out, this was exactly it! I added:</p>
<pre><code class="lang-auto">with:
  token: ${{ secrets.GH_TOKEN }}
</code></pre>
<p>to <code>actions/checkout@v3</code> and it solved the issue. Still not sure why I didn’t need that on other repos, but I will update them all to do this now that I know it was not using the PAT.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="nickderobertis" data-post="4" data-topic="248028">
<div class="title">
<div class="quote-controls"></div>
 nickderobertis:</div>
<blockquote>
<p>Still not sure why I didn’t need that on other repos</p>
</blockquote>
</aside>
<p>I can’t be sure, obviously, but the first thing that comes to mind: In the repository settings you can configure whether the GITHUB_TOKEN should have read-write or read-only access. Maybe that’s different between the repositories?</p>

**Comment by nickderobertis** [comment]
<blockquote>
<p>In the repository settings you can configure whether the GITHUB_TOKEN should have read-write or read-only access. Maybe that’s different between the repositories?</p>
</blockquote>
<p>Ah, yes, that was the underlying reason. I have no idea how this setting got set differently on the repos as I haven’t touched it. But good to know, thanks so much for your help!</p>

**Comment by Dj2chris** [comment]
<p>Why does there have to be so many links</p>

**Comment by yusukekuro** [comment]
<p>Check your workflow permission in <code>Settings &gt; Actions &gt; General</code> .<br>
To push a commit, workflow needs to have write permission.</p><aside class="onebox githubissue" data-onebox-src="https://github.com/ad-m/github-push-action/issues/96#issuecomment-889984928">
  <header class="source">

      <a href="https://github.com/ad-m/github-push-action/issues/96#issuecomment-889984928" target="_blank" rel="noopener nofollow ugc">github.com/ad-m/github-push-action</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/ad-m/github-push-action/issues/96#issuecomment-889984928" target="_blank" rel="noopener nofollow ugc">remote: Permission to git denied to github-actions[bot].</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-07-29" data-time="18:27:02" data-timezone="UTC">06:27PM - 29 Jul 21 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/Shaquu" target="_blank" rel="noopener nofollow ugc">
          <img alt="Shaquu" src="https://user-images.githubusercontent.com/40115800/181099273-86bc3c76-709d-450f-9d9f-1a5b40eef615.jpeg" class="onebox-avatar-inline" width="20" height="20">
          Shaquu
        </a>
      </div>
    </div>

    <div class="labels">
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">Hi,
I am trying to push changes but like you will see it fails due to some erro<span class="show-more-container"><a href="" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">rs.

Test run is here:
https://github.com/NRCHKB/node-red-contrib-homekit-docker/runs/3194895671?check_suite_focus=true

Error
```
Run ad-m/github-push-action@master
  with:
    github_token: ***
    branch: refs/pull/42/merge
    directory: .
Push to branch refs/pull/42/merge
remote: Permission to NRCHKB/node-red-contrib-homekit-docker.git denied to github-actions[bot].
fatal: unable to access 'https://github.com/NRCHKB/node-red-contrib-homekit-docker.git/': The requested URL returned error: 403
Error: Invalid exit code: 128
    at ChildProcess.&lt;anonymous&gt; (/home/runner/work/_actions/ad-m/github-push-action/master/start.js:29:21)
    at ChildProcess.emit (events.js:210:5)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5) {
  code: 128
}
Error: Invalid exit code: 128
    at ChildProcess.&lt;anonymous&gt; (/home/runner/work/_actions/ad-m/github-push-action/master/start.js:29:21)
    at ChildProcess.emit (events.js:210:5)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
```

Part of the gh action:
```yml
  dependabot:
    #needs: [build]
    runs-on: ubuntu-latest
    if: ${{ github.actor == 'dependabot[bot]' }}
    steps:
      - uses: actions/setup-node@v2
        with:
          node-version: '14'
      - uses: actions/checkout@v2
        with:
          persist-credentials: false
          fetch-depth: 0
      - run: |
          git config user.name github-actions[bot]
          git config user.email github-actions[bot]@users.noreply.github.com
          npm version patch -m "[RELEASE] %s"
      - name: Push changes
        uses: ad-m/github-push-action@master
        with:
          github_token: ${{ secrets.GITHUB_TOKEN }}
          branch: ${{ github.ref }}
      - uses: fastify/github-action-merge-dependabot@v2.1.1
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
```</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by yondr-agency-15** [comment]
thanks to @yusukekuro, I found  this option in the settings of my repo: `Settings > Actions > General`

![image](https://user-images.githubusercontent.com/93581121/230569374-078ed259-1bd8-4245-82fe-fb0632e140c4.png)

But adding `permissions` to my .yml also worked, and I prefer it this way:
```
permissions:
  deployments: write
  contents: write
  statuses: write
  actions: write
  checks: read
```
[Github permissions documentation](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idpermissions)

**Reply by nikk30-lilly** [reply]
the solution  working for me

**Reply by pz-max** [reply]
Yes, adding only the permission information to the deploy.yml helped
while keeping the organization "Read and Write permission" turned off

**Comment by martimpassos** [comment]
I'm facing the same error, however in my case I'm checking out a main repo with a submodule (using checkout@v4), then in another step making changes to this submodule and trying to push them. Here's the step:
```
- name: Commit and push changes
        env:
          PERSONAL_ACCESS_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
        run: |
          cd my-submodule
          git config user.name my-username
          git config user.email my-username@github.com
          git remote set-url origin https://x-access-token:${PERSONAL_ACCESS_TOKEN}@github.com/organization/submodule-repo.git
          git add .
          git commit -m "Auto updated data" && git push origin HEAD:main || echo "No changes to commit"
          cd ..
          git add my-submodule
          git commit -m "Auto updated submodule reference" && git push || echo "No changes to commit"
```

Any leads here?

**Reply by airtower-luna** [reply]
`actions/checkout` configures Git to use the token it uses in the `Authorization` header (see [its readme](https://github.com/actions/checkout?tab=readme-ov-file#checkout-v4)), this takes priority over anything you put into the URL (which isn't good style anyway because the URL could leak more easily).

Try setting the authorization header in the submodule Git config instead.

**Comment by martimpassos** [comment]
Based on your answer I simply configured `actions/checkout` to use the PAT and it worked. I didn't realize the credentials persisted between steps, messing with `persist-credentials` could probably work too
